src/odemis/driver/actuator.py

```diff
@@ -2179,3 +2182,516 @@ def terminate(self):
+        :param role: (string)
+        :param parent (Component): the parent of this component, that will be in
+        .parent (the LinkedHeightActuator object)
+        :param dependencies (dict str -> actuator): axis name (in this
+        actuator) -> actuator to be used for this axis (Presumably the lens Z axis)
+        :param rng: the focus range
+        """
+        self.parent = parent
+        self.range = rng
+        axes_def = {
+            "z": model.Axis(unit="m", range=rng),
+        }
+
+        model.Actuator.__init__(self, name, role, parent=parent, axes=axes_def, **kwargs)
+        self._metadata[model.MD_HW_NAME] = "LinkedHeightFocus"
+        # Get the shared executor from the parent
+        self._executor = parent.executor
+
+        self.position = model.VigilantAttribute({},
+                                                unit="m", readonly=True)
+
+        self.updateMetadata({model.MD_POS_COR: -0.005})
+        # Get the underlying lens stage from the passed dependencies list
+        self._lensz = list(dependencies.values())[0]
+        # position will be updated whenever the underlying lens stage is changed
+        self._lensz.position.subscribe(self.updatePosition)
+        self.updatePosition(self._lensz.position.value)
+
+        # Set the focus deactive value from the following special case:
+        # (when the lens is in deactive mode & parent stage Z is at max)
+        focus_value = self._getFocusValue(target_lensz=self._lensz.getMetadata()[model.MD_FAV_POS_DEACTIVE],
+                                          target_stagez=self.parent.axes['z'].range[1])
+        self.updateMetadata({model.MD_FAV_POS_DEACTIVE: focus_value})
+        # If the current initial position is not in focus, set md active position with focus range average
+        if not self.isInFocus():
+            self.updateMetadata({model.MD_FAV_POS_ACTIVE: numpy.mean(self.range)})
+
+    def updateMetadata(self, md):
+        self._metadata.update(md)
+
+    def updatePosition(self, lens_pos):
+        """
+        update the position VA from the underlying dependency
+        (the method is public to be used from the parent)
+        """
+        if 'z' not in lens_pos:
+            return
+        focus_val = self._getFocusValue(target_lensz=lens_pos['z'])
+        logging.debug("reporting position %s", focus_val)
+        self.position._set_value({'z': focus_val}, force_write=True)
+        # If the new position is in focus range, set the MD_FAV_POS_ACTIVE with this new value
+        if self.isInFocus():
+            self.updateMetadata({model.MD_FAV_POS_ACTIVE: self.position.value['z']})
+
+    def isInFocus(self):
+        """
+        A helper function to check if current position is in focus range
+        (the method is public to be used from the parent)
+        :return: True if position in focus range, None otherwise
+        """
+        if self.range[0] <= self.position.value['z'] <= self.range[1]:
```

Comment:
Is there a case where we need to deal with position being greater than the max range? @pieleric

Reply:
So if I got you right, marking it as out of range would be enough, or it needs more special handling? @pieleric